[TF2] Implement tf_stats_bogus ConVars to allow High stat tracking in modes with Increased Stats. I.E MvM

ChaseTownsend opened the pull request:
Implements some Convars to let Server owners/operators increase the maximum stat values to accept or always track stats

`tf_stats_bogus_damage_max`: The maximum damage allowed before a bogus warning is sent. <br> Default: `1500`

`tf_stats_bogus_damage_mvm_max`: The maximum damage allowed in MvM before a bogus warning is sent. <br> Default: `5000`

`tf_stats_bogus_healing_max`: The maximum healing allowed before a bogus warning is sent. <br> Default: `1000`

`tf_stats_bogus_block_damage_max`: The maximum damage blocked allowed before a bogus warning is sent. <br> Default: `3000`

`tf_stats_bogus_ignore`: If non-zero, ignore recording bogus stats. If equals 2, don't print DevMsg. <br> Default: `1` "Yes/Ignore bogus + DevMsg" <br> Minimum: `0` Maximum: `2`

ChaseTownsend committed:
Implement ConVars

ZzZombo commented on src/game/server/tf/tf_gamestats.cpp:
```diff
@@ -40,6 +40,11 @@ extern ConVar tf_mm_trusted;
 static ConVar tf_stats_nogameplaycheck( "tf_stats_nogameplaycheck", "0", FCVAR_NONE , "Disable normal check for valid gameplay, send stats regardless." );
 //static ConVar tf_stats_track( "tf_stats_track", "1", FCVAR_NONE, "Turn on//off tf stats tracking." );
 //static ConVar tf_stats_verbose( "tf_stats_verbose", "0", FCVAR_NONE, "Turn on//off verbose logging of stats." );
+ConVar tf_stats_bogus_damage_max("tf_stats_bogus_damage_max", "1500", FCVAR_REPLICATED, "Maximum Damage before Bogus warning");
```
Please fix up the spelling and grammar here and everywhere else in the PR. The stuff like inconsistent capitalization is very jarring.

ChaseTownsend replied:
I will fix up the info in the convars. Are there any others that are not the convars that I missed?

ChaseTownsend replied:
Fixed the inconsistent capitalization. If you specify any grammar that could be reworded or changed, let me know.

ChaseTownsend replied:
Just fixed that. Thanks for your feedback!

ZzZombo replied:
Sorry, I am stupid. "MVM" should be "MvM". Sorry about this.

ChaseTownsend committed:
Fix multiple spelling mistakes

ChaseTownsend committed:
Replace "mvm" in convar msg to "MVM"

ChaseTownsend committed:
Replace "MVM" with "MvM"

ChaseTownsend committed:
Combine an old if statement

ChaseTownsend committed:
Fix changes made. Remove whitespace

Fixed an if statement, and remove white space

ChaseTownsend commented on src/game/server/tf/tf_gamestats.cpp:
```diff
@@ -1033,14 +1060,25 @@ void CTFGameStats::Event_PlayerFiredWeapon( CTFPlayer *pPlayer, bool bCritical )
-	const int INSANE_PLAYER_DAMAGE = TFGameRules()->IsMannVsMachineMode() ? 5000 : 1500;
-
+	int INSANE_PLAYER_DAMAGE = TFGameRules()->IsMannVsMachineMode() ? tf_stats_bogus_damage_mvm_max.GetFloat() : tf_stats_bogus_damage_max.GetFloat();
+	
```
Fixed Whitespace.

ChaseTownsend committed:
Replace .GetFloat() with .GetInt() when storing into an Int

ChaseTownsend commented:
Drafting this until I do the Changes specified by FlaminSarge

ChaseTownsend committed:
Merge branch 'ValveSoftware:master' into Stat-Tracking-CVars

ChaseTownsend committed:
New ConVar changes

Implement FlaminSarges sugestions

ChaseTownsend commented:
@FlaminSarge Updated the PR info and convars

ChaseTownsend committed:
general fixes from prev commit

revert converting some `const int` to `int`
fix maximum damage convars getting floats and storing into an int

ChaseTownsend committed:
Merge branch 'ValveSoftware:master' into Stat-Tracking-CVars

ChaseTownsend committed:
Fix issues

ChaseTownsend committed:
whitespace fix

ChaseTownsend committed:
Fix more Whitespace

FlaminSarge commented on src/game/server/tf/tf_gamestats.cpp:
```diff
@@ -40,6 +40,11 @@ extern ConVar tf_mm_trusted;
+ConVar tf_stats_bogus_damage_mvm_max("tf_stats_bogus_damage_mvm_max", "5000", FCVAR_REPLICATED, "The maximum damage allowed in mvm before a bogus warning is sent.", true, 5000.f, false, 0.f );
+ConVar tf_stats_bogus_healing_max("tf_stats_bogus_healing_max", "1000", FCVAR_REPLICATED, "The maximum healing allowed before a bogus warning is sent.", true, 1000.f, false, 0.f );
+ConVar tf_stats_bogus_block_damage_max("tf_stats_bogus_block_damage_max", "3000", FCVAR_REPLICATED, "The maximum damage blocked allowed before a bogus warning is sent.", true, 3000.f, false, 0.f );
+ConVar tf_stats_bogus_ignore("tf_stats_bogus_ignore", "1", FCVAR_REPLICATED, "If non-zero, ignore recording bogus stats. If equals 2, don't print DevMsg.", true, 0.f, true, 2.f );
```
For what reason is the _minimum_ value of the cvar set to the default? While atypical, why not let users set the cvars lower?

ChaseTownsend replied:
Not much of a real point because the only use case would be if you wanted it to not track these stats at all

FlaminSarge replied:
(That is a real use case).

FlaminSarge commented on src/game/server/tf/tf_gamestats.cpp:
```diff
@@ -1261,11 +1287,18 @@ void CTFGameStats::Event_PlayerDamageAssist( CBasePlayer *pProvider, int iBonusD
 //-----------------------------------------------------------------------------
 void CTFGameStats::Event_BossDamage( CBasePlayer *pAttacker, int iDamage )
 {
-	const int INSANE_DAMAGE = 5000;
+	const int INSANE_DAMAGE = tf_stats_bogus_damage_mvm_max.GetInt();
 	Assert( iDamage >= 0 );
 	Assert( iDamage <= INSANE_DAMAGE );
-	if ( iDamage < 0 || iDamage > INSANE_DAMAGE )
+
+	if ( ( iDamage < 0 || iDamage > INSANE_DAMAGE ) && tf_stats_bogus_ignore.GetInt() > 0 )
+	{
```
This seems like it needs some changes to avoid assert-failing if the cvar is set super low.
Also not sure (can't remember) if you can set the const value to the cvar value; better to just use the cvar value where needed instead (only one place since you'd be removing the assert).

ChaseTownsend replied:
should this go for the others that were also modified?
such as `Event_PlayerDamageAssist` line 1262

FlaminSarge replied:
I'm honestly not sure. In dev, the Assert will stop the game entirely, but on the production build it may(?) not have any effect other than noise. It doesn't seem right to assert on the bogus values now that they're configurable, though.

ChaseTownsend replied:
should they be commented out or just removed

ChaseTownsend committed:
Remove Cvar Limits

ChaseTownsend committed:
Remove old Asserts

ChaseTownsend committed:
Merge branch 'ValveSoftware:master' into Stat-Tracking-CVars
